[MM-12649] Convert spreadsheet cells to markdown table when pasting

amorriscode opened the pull request:
#### Summary
This PR resolves [#9759](https://github.com/mattermost/mattermost-server/issues/9759). Users can now paste copied sheets (from Google Sheets or Excel) and they will be formatted to Markdown. I based it heavily off of GitHub's [paste-markdown](https://github.com/github/paste-markdown) library but could not use the library directly.

#### Ticket Link
[Jira ticket](https://mattermost.atlassian.net/browse/MM-12649) | [#9759](https://github.com/mattermost/mattermost-server/issues/9759)

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Touches critical sections of the codebase (posting)

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0c27e6dd553903275.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0c27e6dd553903275

cpanato commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

cpanato commented:
Spintmint test server creation failed. See the logs for more information.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

amorriscode commented:
@esethna I apologize for all the back and forth.

I needed to format tables using GitHub Markdown rules so a single column table works now!

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0536d38350d7465c3.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0536d38350d7465c3

hanzei commented:
@esethna I set up a new test server with the latest changes. Could you give this another look?

mattermod commented:
Spinmint test server destroyed

hanzei commented:
@hmhealey @saturninoabril Can you please take a look? It would be great, if we got this in v5.8.0

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-08ed469a934da65fb.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-08ed469a934da65fb

hmhealey commented on utils/post_utils.jsx:
```diff
@@ -216,3 +216,54 @@ export function isErrorInvalidSlashCommand(error) {
     return false;
 }
+
+function parseTable(html) {
```
Perhaps we could move all of these functions into their own file named something like `utils/paste.js`? They're self-contained, so it might be nice to keep them on their own.

amorriscode commented:
I'll work on moving the code around, renaming, and adding some unit tests.

In regards to the first row being a header, I don't know if we can get around that. It seems like most Markdown parsers don't support headerless tables (including GitHub).

amorriscode commented:
@saturninoabril @hmhealey I implemented the changes you both requested. Added some better tests and brought out the paste functions into their own utils file. Let me know if there is anything else you'd like me to change! 😀

saturninoabril commented on utils/paste.test.jsx:
```diff
@@ -0,0 +1,55 @@
+            expect(formatMarkdownTableMessage(validTable, testMessage)).toBe(`${testMessage}\n\n${markdownTable}`);
+        });
+    });
+});
```
💯

hmhealey commented:
As for the first row being the header, that's correct that it's just part of Markdown. It would be nice if we had some sort of control over that, but we keep having problems every time we try to deviate from the Markdown spec.

mattermod commented:
Spinmint test server destroyed
